### ISSUE TITLE
Věci nutné k dotažení pro vypublikování do beta režimu

### ISSUE THREAD

**Issue by padak** [keboola/app-snowflake-dwh-manager#12]
# beta verze

Aplikaci dáváme do public režimu, v popisku je napsané, že potřebuje od nás dostat DB. Pokud to dává smysl, tak popsal, jak to udělat, aby to fungovalo na klientově Snowflake.

## hesla

- založený user by měl dostat email na kterém si vyzvedne heslo, případně by to init heslo mělo být vidět v cfg (když se po zalogování stejně musí měnit) - nemělo by to vyžadovat lovit ho v eventech
- heslo by mělo jít v manageru změnit (v Mallu useři svoje hesla už stihli zapomenout)

# verze 1

Aplikace dělá provisioning Snowflake DB (nepotřebuješ support) a můžeš tak spravovat víc DBs. V Mallu tak například chtějí druhou DB pro Tableau, aby mohl k datům Tableau i zaměstnanci, ale mohli je vyhazovat - tohle je use-case Liftaga, co psal dneska Kačurez.

# verze 2

Aplikace se stará i o loadování dat, nemusí se na to myslet a pouštět o Writer + tuhle apku. Prezentujeme ji v hlavním menu pod položkou "Analytical DWH".

**Comment by tomasfejfar** [comment]
V1: pokud poběží pořád jako docker komponenta, tak to bude znamenat, že bude potřeba do ní zahardcodovat přihlašovací údaje, asi. Nebo jí nějak "alternativně" nagenerovat loginy někam třeba nějakým processorem. Nebo rozšířit docker-runner tak, aby to nějak přihazoval do configu než pustí komponentu. 

V2: Neumím si moc představit, jak to udělat jinak, než nějakou ad-hoc úpravou systému - kdy si komponenty začnou povídat mezi sebou. Nebo vznikne nová "higher order component", která bude moct přistupovat k projektu a koukat co se v něm děje. Takový event-driven orchestrátor. Nebo by to mohlo fungovat jako trigger. Založil bys trigger na "něco" (třeba: "proběhne job configu #32145") a na něj navěsíš další job (typicky orchestrátor, který spustí po sobě writer a appku). 

V3: Tohle všechno znamená, že už to není jen komponenta, ale je to zarostlé do KBC a je to na úrovni orchestrátorů a transformací. 

V4: Jakmile bude V1, tak není problém to dělat na kterémkoli SFLK - jen nám musí (nějak) předat nějaké loginy, které má práva vytvářet DB.

**Comment by Halama** [comment]
jo v2 a v3 už z největší pravděpodobností znamená že to bude feature KBC nebo to minimálně bude vyžadovat nové features KBC. Nechtěli jsme to tam cpát rovnou proto se to jako POC udělalo standardní docker-runner komponenta.  Hodil jsem včera jako téma na offsite https://www.wunderlist.com/#/lists/348246733

beta a v1 bysme měli být schopný nějak nahodit nad současným dwhm.

**Comment by padak** [comment]
Ještě mě napadá k V1, že by ty další usecases mohli bejt v tý existující DB, takže v případě Mallu by si pro Tableau udělali druhé `schema` a v managementu uživatelů řešili jestli user X má přístup do `DATAMARTS` nebo `TABLEAU`. Kdyby naše UI na založení userů prošlo Jackem a Janem, tak by to asi bylo praktičtější než jim dávat další DB, ne?

(pro kontext, dneska to mají tady https://connection.eu-central-1.keboola.com/admin/projects/196/applications/keboola.app-snowflake-dwh-manager/4121312)

**Comment by Halama** [comment]
V1, ta appka podporuje více schémat i přiřazení jednotlivých schémat userům.

**Comment by Halama** [comment]
Mě to UI přijde docela fajn. Co se ti tam nezdá. Tady je to žřejmý že můžeš vybrat více schémat. Jediné co imho chybí je nápověda tech schémat.

![image](https://user-images.githubusercontent.com/903531/44139946-ec2ab6fc-a078-11e8-9ea6-776db819a68a.png)

**Comment by padak** [comment]
Myslím že jen úplný drobnosti. Třeba když přidám druhý schéma a přiřadím ho svýmu userovi, tak nejsem schopnej ofiltrovat uživatele podle toho k čemu maj přístup. Úplně vidím, jak se to v tomhle bude blbě spravovat, když tam dáš desítky uživatelů (k tomu se chystají). Pokud by do toho nahodili management writerů (vlastně by dělali provisioning svejch věcí (schemat)) přes tohle a pak měli jen Writery, jako loadovadla dat, tak by to bylo plný spousty Schemat a k nim různejm (asi jednotky) lidí. Fakticky v tom uživatelé všechno udělají, ale prakticky s tím budou trochu trpět. Myslím, že jde jen o kosmetický UX detaily.

**Comment by ondrejhlavacek** [comment]
Jasně, config rows UI jsme schopný ohnout jen do nějaký míry, pokud to má bejt jinak zaintegrovaný, možná to bude chtít vyjmout z config rows ui (kde to je teď spíš jako poc).

**Comment by ondrejhlavacek** [comment]
ad hesla - to je ještě řešitelný se současným setupem - zasílání hesla mailem a reset hesla přes sync akci (to UI ještě neumí, ale bude umět)

ad decomissioning - to by možná šlo vyřešit taky. musel by se udělat hook, kterej při mazání řádku konfigurace zavolá nějakou sync akci

**Comment by tomasfejfar** [comment]
Decomissioning jde už teď, i když chápu, že to není ideální - stačí toho uživatele dát disabled. Od té chvíle se nemůže přihlásit a je to více méně ekvivalentní smazání - jediná nevýhoda je, že zůstává v seznamu userů.

**Comment by padak** [comment]
Teď jsem tam narazil ještě na pořadí. Fixně by bylo fajn, aby SCHEMATA byly dřív nez USERS, jinak se stane, že se zpracovává user, ale neexistuje pro něj SCHEMA. Stačí pak jen změnit pořadí a jede to OK!

Přidávám to nahoru do popisku ISSUE, ať se to nemusí lovit v komentářích.

(příklad chyby pak tady https://connection.eu-central-1.keboola.com/admin/projects/196/jobs/6794225)

**Comment by ondrejhlavacek** [comment]
Jo, tak tohle je přesně věc, kterou config rows ui neumí a patrně ani umět nebude. Na to bude potřeba napsat vlastní UI.

**Comment by tomasfejfar** [comment]
ad 1: https://github.com/keboola/app-snowflake-dwh-manager/issues/13

**Comment by tomasfejfar** [comment]
[ad 2](https://github.com/keboola/app-snowflake-dwh-manager/issues/12#issuecomment-416025397): Už není potřeba s `GRANT ... TO ALL FUTURE... ` viz #17
